atlas/packages/components/src/components/Carousel/Carousel.tsx

```diff
@@ -1,39 +1,105 @@
+import { animated, useSpring } from "react-spring";
+import useResizeObserver from "use-resize-observer";
 import { useCSS, CarouselStyleProps } from "./Carousel.style";
 import NavButton from "../NavButton";
 
 type CarouselProps = {
-	children: Array<ReactNode>;
-	scrollAmount?: Number;
+	children: React.ReactNode;
+	containerCss: SerializedStyles;
+	leftControlCss: SerializedStyles;
+	rightControlCss: SerializedStyles;
+	onScroll: (direction: "left" | "right") => void;
 } & CarouselStyleProps;
 
-export default function Carousel({ children, scrollAmount = 200, ...styleProps }: CarouselProps) {
-	const container = useRef(null);
-
-	function onScroll(direction: "right" | "left") {
-		container.current.scrollBy({
-			left: direction === "left" ? -scrollAmount : scrollAmount,
-			behavior: "smooth",
-		});
+const Carousel: React.FC<Partial<CarouselProps>> = ({
+	children,
+	containerCss,
+	leftControlCss,
+	rightControlCss,
+	onScroll = () => {},
+}) => {
+	if (!Array.isArray(children)) {
+		return <>{children}</>;
 	}
+	let [props, set] = useSpring(() => ({
+		transform: `translateX(0px)`,
+	}));
+	const [x, setX] = useState(0);
+	const { width: containerWidth, ref: containerRef } = useResizeObserver<HTMLDivElement>();
+	const elementsRefs = useRef<(HTMLDivElement | null)[]>([]);
+	const [childrenLength, setChildrenLength] = useState(0);
+	useEffect(() => {
+		elementsRefs.current = elementsRefs.current.slice(0, children.length);
+		const childrensLength = elementsRefs.current.reduce(
+			(accWidth, el) => (el != null ? accWidth + el.clientWidth : accWidth),
+			0
+		);
+		setChildrenLength(childrensLength);
+	}, [children.length]);
 
-	let styles = useCSS(styleProps);
-
+	const styles = useMemo(() => useCSS({}), []);
 	return (
-		<div css={styles.wrapper}>
-			<div ref={container} css={styles.container}>
-				{children.map((item, index) => (
-					<div key={`carousel-${index}`} css={styles.item}>
-						{item}
-					</div>
+		<div css={[styles.container, containerCss]}>
+			<div css={styles.itemsContainer} ref={containerRef}>
+				{children.map((element, idx) => (
+					<animated.div
+						style={props}
+						key={`Carousel-${idx}`}
+						ref={(el) => {
+							elementsRefs.current[idx] = el;
+							return el;
+						}}
+					>
+						{element}
+					</animated.div>
 				))}
 			</div>
-			<div css={styles.navLeft}>
-				<NavButton type="primary" direction="left" onClick={() => onScroll("left")} />
-			</div>
-			<div css={styles.navRight}>
-				<NavButton type="primary" direction="right" onClick={() => onScroll("right")} />
-			</div>
+			<NavButton
+				outerCss={[styles.navLeft, leftControlCss]}
+				direction="left"
+				onClick={() => {
+					handleScroll("left");
+				}}
+			/>
+			<NavButton
+				outerCss={[styles.navRight, rightControlCss]}
+				direction="right"
+				onClick={() => {
+					handleScroll("right");
+				}}
+			/>
 		</div>
 	);
-}
+
+	function handleScroll(direction: "left" | "right") {
```

Comment:
This should be defined before the `return`. It's a good practice to always have the `return` statement be the last one in a function component.

Reply:
Moved it up. I like to hide functions under return to hide the implementation details away, but this one is quite short, so I moved it up.